File: bci/api/python/wia_bci/core/wia_bci.py
"""WIA BCI Main Class."""

import logging
from typing import Any, Callable, Optional

from wia_bci.core.event_emitter import BciEventEmitter
from wia_bci.core.bci_error import BciError
from wia_bci.types.config import DeviceConfig, LogLevel, WiaBciOptions
from wia_bci.types.device import BciState, ChannelInfo, DeviceInfo
from wia_bci.types.events import EventType, SignalEvent, ErrorEvent

from wia_bci.adapters.base_adapter import BaseBciAdapter
from wia_bci.adapters.simulator_adapter import SimulatorAdapter


def _create_logger(level: LogLevel) -> logging.Logger:
    """Create logger with specified level."""
    logger = logging.getLogger("wia-bci")

    level_map = {
        "debug": logging.DEBUG,
        "info": logging.INFO,
        "warn": logging.WARNING,
        "error": logging.ERROR,
    }

    logger.setLevel(level_map.get(level, logging.INFO))

    if not logger.handlers:
        handler = logging.StreamHandler()
        handler.setFormatter(
            logging.Formatter("[WIA-BCI] [%(levelname)s] %(message)s")
        )
        logger.addHandler(handler)

    return logger


class WiaBci:
    """
    WiaBci - Main BCI Interface Class.

    Example:
        >>> bci = WiaBci()
        >>> await bci.connect(DeviceConfig(type='eeg_headset'))
        >>> bci.on('signal', lambda e: print(e.data))
        >>> await bci.start_stream()
    """

    def __init__(self, options: Optional[WiaBciOptions] = None) -> None:
        self._options = options or WiaBciOptions()
        self._emitter = BciEventEmitter()
        self._logger = _create_logger(self._options.log_level)
        self._adapter: Optional[BaseBciAdapter] = None
        self._connected = False
        self._streaming = False
        self._samples_received = 0
        self._reconnect_attempts = 0

        self._logger.debug(f"WiaBci initialized with options: {self._options}")

    async def list_devices(self) -> list[DeviceInfo]:
        """List available BCI devices."""
        self._logger.debug("Listing devices...")

        from wia_bci.types.device import DeviceCapabilities

        # Return simulator as available device
        devices = [
            DeviceInfo(
                id="simulator-001",
                name="WIA BCI Simulator",
                type="simulator",
                manufacturer="WIA",
                model="Simulator",
                capabilities=DeviceCapabilities(
                    channels=8,
                    max_sampling_rate=250,
                    supported_sampling_rates=[125, 250],
                    resolution=24,
                ),
                status="available",
            )
        ]

        return devices

    async def connect(self, config: DeviceConfig) -> None:
        """Connect to a BCI device."""
        if self._connected:
            raise BciError.already_connected()

        self._logger.info(f"Connecting to device: {config}")

        try:
            self._adapter = self._create_adapter(config)
            await self._adapter.connect(config)

            # Set up handlers
            def on_data(event: SignalEvent) -> None:
                self._samples_received += 1
                self._emitter.emit("signal", event)
                self._emitter.emit("data", event)

            def on_error(error: ErrorEvent) -> None:
                self._logger.error(f"Adapter error: {error}")
                self._emitter.emit("error", error)

            self._adapter.on_data(on_data)
            self._adapter.on_error(on_error)

            self._connected = True
            self._reconnect_attempts = 0

            self._logger.info("Connected successfully")
            self._emitter.emit("connected")

        except BciError:
            raise
        except Exception as e:
            self._logger.error(f"Connection failed: {e}")
            raise BciError.connection_failed(str(e), e)

    async def disconnect(self) -> None:
        """Disconnect from the current device."""
        if not self._connected or not self._adapter:
            raise BciError.not_connected()

        self._logger.info("Disconnecting...")

        if self._streaming:
            await self.stop_stream()

        await self._adapter.disconnect()
        self._adapter = None
        self._connected = False

        self._logger.info("Disconnected")
        self._emitter.emit("disconnected")

    def is_connected(self) -> bool:
        """Check if connected."""
        return self._connected

    async def start_stream(self) -> None:
        """Start data streaming."""
        if not self._connected or not self._adapter:
            raise BciError.not_connected()

        if self._streaming:
            self._logger.warning("Already streaming")
            return

        self._logger.info("Starting stream...")

        await self._adapter.start_stream()
        self._streaming = True

        self._logger.info("Stream started")
        self._emitter.emit("stream_started")

    async def stop_stream(self) -> None:
        """Stop data streaming."""
        if not self._connected or not self._adapter:
            raise BciError.not_connected()

        if not self._streaming:
            self._logger.warning("Not streaming")
            return

        self._logger.info("Stopping stream...")

        await self._adapter.stop_stream()
        self._streaming = False

        self._logger.info("Stream stopped")
        self._emitter.emit("stream_stopped")

    def is_streaming(self) -> bool:
        """Check if streaming."""
        return self._streaming

    def get_device_info(self) -> Optional[DeviceInfo]:
        """Get current device info."""
        if self._adapter:
            return self._adapter.get_device_info()
        return None

    def get_channels(self) -> list[ChannelInfo]:
        """Get channel information."""
        if self._adapter:
            return self._adapter.get_channels()
        return []

    def get_state(self) -> BciState:
        """Get current BCI state."""
        device_info = self.get_device_info()
        return BciState(
            connected=self._connected,
            streaming=self._streaming,
            device=device_info,
            channels=self.get_channels(),
            sampling_rate=(
                device_info.capabilities.max_sampling_rate if device_info else 0
            ),
            samples_received=self._samples_received,
            last_error=None,
        )

    def on(self, event: EventType, handler: Callable[..., None]) -> "WiaBci":
        """Subscribe to an event."""
        self._emitter.on(event, handler)
        return self

    def off(self, event: EventType, handler: Callable[..., None]) -> "WiaBci":
        """Unsubscribe from an event."""
        self._emitter.off(event, handler)
        return self

    def once(self, event: EventType, handler: Callable[..., None]) -> "WiaBci":
        """Subscribe to an event once."""
        self._emitter.once(event, handler)
        return self

    def emit(self, event: EventType, data: Optional[Any] = None) -> "WiaBci":
        """Emit an event (for external use, e.g., markers)."""
        self._emitter.emit(event, data)
        return self

    def dispose(self) -> None:
        """Clean up resources."""
        self._logger.debug("Disposing WiaBci...")

        if self._adapter:
            self._adapter.dispose()
            self._adapter = None

        self._emitter.remove_all_listeners()
        self._connected = False
        self._streaming = False

        self._logger.debug("Disposed")

    def _create_adapter(self, config: DeviceConfig) -> BaseBciAdapter:
        """Create appropriate adapter for device type."""
        if config.type == "simulator":
            return SimulatorAdapter()

        if config.type in ("eeg_headset", "eeg_cap"):
            self._logger.warning(
                f"Using simulator for device type: {config.type}"
            )
            return SimulatorAdapter()

        raise BciError.invalid_config(f"Unsupported device type: {config.type}")

File: bci/api/typescript/src/core/WiaBci.ts
/**
 * WIA BCI Main Class
 * @module wia-bci/core/WiaBci
 */

import { BciEventEmitter } from './EventEmitter';
import { BciError, ErrorCodes } from './BciError';
import type {
  WiaBciOptions,
  DeviceConfig,
  DeviceInfo,
  ChannelInfo,
  BciState,
  EventType,
  EventHandler,
  SignalEvent,
  DEFAULT_OPTIONS,
} from '../types';
import type { IBciAdapter } from '../adapters/BaseAdapter';
import { SimulatorAdapter } from '../adapters/SimulatorAdapter';

/**
 * Logger interface
 */
interface Logger {
  debug: (message: string, ...args: unknown[]) => void;
  info: (message: string, ...args: unknown[]) => void;
  warn: (message: string, ...args: unknown[]) => void;
  error: (message: string, ...args: unknown[]) => void;
}

/**
 * Create logger based on log level
 */
function createLogger(level: WiaBciOptions['logLevel']): Logger {
  const levels = ['debug', 'info', 'warn', 'error'];
  const minLevel = levels.indexOf(level ?? 'info');

  const log = (logLevel: string, message: string, ...args: unknown[]) => {
    if (levels.indexOf(logLevel) >= minLevel) {
      const prefix = `[WIA-BCI] [${logLevel.toUpperCase()}]`;
      console[logLevel as 'log'](prefix, message, ...args);
    }
  };

  return {
    debug: (message, ...args) => log('debug', message, ...args),
    info: (message, ...args) => log('info', message, ...args),
    warn: (message, ...args) => log('warn', message, ...args),
    error: (message, ...args) => log('error', message, ...args),
  };
}

/**
 * WiaBci - Main BCI Interface Class
 *
 * @example
 * ```typescript
 * const bci = new WiaBci();
 *
 * await bci.connect({ type: 'eeg_headset' });
 *
 * bci.on('signal', (event) => {
 *   console.log('Signal:', event.data);
 * });
 *
 * await bci.startStream();
 * ```
 */
export class WiaBci {
  private emitter: BciEventEmitter;
  private logger: Logger;
  private options: Required<WiaBciOptions>;
  private adapter: IBciAdapter | null = null;
  private _connected = false;
  private _streaming = false;
  private samplesReceived = 0;
  private reconnectAttempts = 0;

  constructor(options?: WiaBciOptions) {
    this.options = {
      autoReconnect: options?.autoReconnect ?? true,
      reconnectInterval: options?.reconnectInterval ?? 3000,
      maxReconnectAttempts: options?.maxReconnectAttempts ?? 5,
      bufferSize: options?.bufferSize ?? 1000,
      logLevel: options?.logLevel ?? 'info',
    };

    this.emitter = new BciEventEmitter();
    this.logger = createLogger(this.options.logLevel);

    this.logger.debug('WiaBci initialized', this.options);
  }

  /**
   * List available BCI devices
   */
  async listDevices(): Promise<DeviceInfo[]> {
    this.logger.debug('Listing devices...');

    // In a real implementation, this would scan for devices
    // For now, return simulator as available
    const devices: DeviceInfo[] = [
      {
        id: 'simulator-001',
        name: 'WIA BCI Simulator',
        type: 'simulator',
        manufacturer: 'WIA',
        model: 'Simulator',
        capabilities: {
          channels: 8,
          maxSamplingRate: 250,
          supportedSamplingRates: [125, 250],
          resolution: 24,
          hasAccelerometer: false,
          hasGyroscope: false,
          hasImpedanceCheck: false,
          hasBatteryIndicator: false,
          supportedProtocols: [],
        },
        status: 'available',
      },
    ];

    return devices;
  }

  /**
   * Connect to a BCI device
   */
  async connect(config: DeviceConfig): Promise<void> {
    if (this._connected) {
      throw BciError.alreadyConnected();
    }

    this.logger.info('Connecting to device...', config);

    try {
      // Select appropriate adapter based on device type
      this.adapter = this.createAdapter(config);

      // Connect via adapter
      await this.adapter.connect(config);

      // Set up data handler
      this.adapter.onData((event) => {
        this.samplesReceived++;
        this.emitter.emit('signal', event);
        this.emitter.emit('data', event);
      });

      // Set up error handler
      this.adapter.onError((error) => {
        this.logger.error('Adapter error:', error);
        this.emitter.emit('error', error);
      });

      this._connected = true;
      this.reconnectAttempts = 0;

      this.logger.info('Connected successfully');
      this.emitter.emit('connected');
    } catch (error) {
      this.logger.error('Connection failed:', error);

      if (error instanceof BciError) {
        throw error;
      }

      throw BciError.connectionFailed(
        error instanceof Error ? error.message : 'Unknown error',
        error
      );
    }
  }

  /**
   * Disconnect from the current device
   */
  async disconnect(): Promise<void> {
    if (!this._connected || !this.adapter) {
      throw BciError.notConnected();
    }

    this.logger.info('Disconnecting...');

    if (this._streaming) {
      await this.stopStream();
    }

    await this.adapter.disconnect();
    this.adapter = null;
    this._connected = false;

    this.logger.info('Disconnected');
    this.emitter.emit('disconnected');
  }

  /**
   * Check if connected
   */
  isConnected(): boolean {
    return this._connected;
  }

  /**
   * Start data streaming
   */
  async startStream(): Promise<void> {
    if (!this._connected || !this.adapter) {
      throw BciError.notConnected();
    }

    if (this._streaming) {
      this.logger.warn('Already streaming');
      return;
    }

    this.logger.info('Starting stream...');

    await this.adapter.startStream();
    this._streaming = true;

    this.logger.info('Stream started');
    this.emitter.emit('stream_started');
  }

  /**
   * Stop data streaming
   */
  async stopStream(): Promise<void> {
    if (!this._connected || !this.adapter) {
      throw BciError.notConnected();
    }

    if (!this._streaming) {
      this.logger.warn('Not streaming');
      return;
    }

    this.logger.info('Stopping stream...');

    await this.adapter.stopStream();
    this._streaming = false;

    this.logger.info('Stream stopped');
    this.emitter.emit('stream_stopped');
  }

  /**
   * Check if streaming
   */
  isStreaming(): boolean {
    return this._streaming;
  }

  /**
   * Get current device info
   */
  getDeviceInfo(): DeviceInfo | null {
    return this.adapter?.getDeviceInfo() ?? null;
  }

  /**
   * Get channel information
   */
  getChannels(): ChannelInfo[] {
    return this.adapter?.getChannels() ?? [];
  }

  /**
   * Get current BCI state
   */
  getState(): BciState {
    return {
      connected: this._connected,
      streaming: this._streaming,
      device: this.getDeviceInfo(),
      channels: this.getChannels(),
      samplingRate: this.adapter?.getDeviceInfo()?.capabilities.maxSamplingRate ?? 0,
      samplesReceived: this.samplesReceived,
      lastError: null,
    };
  }

  /**
   * Subscribe to an event
   */
  on<T extends EventType>(event: T, handler: EventHandler<T>): this {
    this.emitter.on(event, handler);
    return this;
  }

  /**
   * Unsubscribe from an event
   */
  off<T extends EventType>(event: T, handler: EventHandler<T>): this {
    this.emitter.off(event, handler);
    return this;
  }

  /**
   * Subscribe to an event once
   */
  once<T extends EventType>(event: T, handler: EventHandler<T>): this {
    this.emitter.once(event, handler);
    return this;
  }

  /**
   * Emit an event (for external use, e.g., markers)
   */
  emit<T extends EventType>(event: T, data?: Parameters<EventHandler<T>>[0]): this {
    this.emitter.emit(event, data as never);
    return this;
  }

  /**
   * Clean up resources
   */
  dispose(): void {
    this.logger.debug('Disposing WiaBci...');

    if (this._connected && this.adapter) {
      this.adapter.dispose();
      this.adapter = null;
    }

    this.emitter.removeAllListeners();
    this._connected = false;
    this._streaming = false;

    this.logger.debug('Disposed');
  }

  /**
   * Create appropriate adapter for device type
   */
  private createAdapter(config: DeviceConfig): IBciAdapter {
    switch (config.type) {
      case 'simulator':
        return new SimulatorAdapter();

      case 'eeg_headset':
      case 'eeg_cap':
        // In production, would check manufacturer and return appropriate adapter
        // For now, use simulator
        this.logger.warn(`Using simulator for device type: ${config.type}`);
        return new SimulatorAdapter();

      default:
        throw BciError.invalidConfig(`Unsupported device type: ${config.type}`);
    }
  }
}
